src/nelder_mead.jl

```diff
@@ -0,0 +1,118 @@
+export nelder_mead
+
+"""
+    nelder_mead(nlp)
+
+This implementation follows the algorithm described in chapter 9 of [1].
+[1] Numerical Optimization (Jorge Nocedal and Stephen J. Wright), Springer, 2006.
+"""
+function nelder_mead(nlp :: AbstractNLPModel;
+                     x :: AbstractVector = copy(nlp.meta.x0),
+                     vertices :: Vector{<: AbstractVector{<: AbstractFloat}} = Vector{eltype(x)}[],
+                     tol :: Real = √eps(eltype(x)),
+                     max_time :: Float64 = 30.0,
+                     max_eval :: Int = -1,
+                     ref_prm :: Real = - one(eltype(x)),
+                     exp_prm :: Real = - one(eltype(x)) * 2,
+                     ocn_prm :: Real = - one(eltype(x)) / 2,
+                     icn_prm :: Real = one(eltype(x)) / 2)
+
+  # Initial simplex
+  n = nlp.meta.nvar
+  T = eltype(x)
+  if n < length(vertices) < 0
+    error("Invalid Simplex : The number of initial points is less than n + 1")
+  elseif length(vertices) == 0
+    vertices = [x,[copy(x) for i = 1:n]...]
+    for j in 1:n
+      xt = vertices[j + 1]
+      xt[j] += xt[j] == 0 ? eps(T)^T(0.25) : T(0.05)
+    end
+  end
+
+  pairs = [(x, obj(nlp, x)) for x in vertices]
+  sort!(pairs, by=x->x[2])
+
+  x_trial = zeros(eltype(x), n)
+  k = 0
+  el_time = 0.0
+  start_time = time()
+  tired = neval_obj(nlp) > max_eval >= 0 || el_time > max_time
+  norm_opt = norm(first(pairs)[1] - last(pairs)[1])
+  optimal = norm_opt < tol
+  T = eltype(x)
+  status =:unknown
+  @info log_header([:iter, :f, :nrm], [Int, T, T], hdr_override=Dict(:f => "f(x)", :nrm => "‖x₁ - xₙ+1‖"))
+
+  while !(optimal || tired)
+
+    shrink = true
+    x_cen = sum(pairs[i][1] for i in 1:n)/T(n)
```

Comment:
The ``` .= ``` assignment here (to compute the reflection point) seems to break the algorithm somehow, i'll have to take a deeper look at it.

Reply:
`.=` only works if the lhs is defined, I think `x_cen` is first defined here? Create it outside the loop. (`zeros(T, n)`).